one-ui/src/main/java/com/marklogic/hub/oneui/controllers/EnvironmentController.java

```diff
@@ -90,31 +96,109 @@ public JsonNode getProject() {
+
+        // delete contents from the current project folder
+        cleanExistingFSProject(project, listener);
+
+        // extract the uploaded & zipped project file into the current project folder
+        extractZipProject(project, uploadedFile, listener);
+
+        HubDeployStatusListener uninstalllistener = getInstallListener(dataHubConfigurationException, true);
+        dataHubService.unInstall(uninstalllistener);
+
+        install(project.getProjectDirString(), listener, dataHubConfigurationException);
+    }
+
+    private HubDeployStatusListener getInstallListener(final Exception[] dataHubConfigurationException, boolean isUninstall) {
         HubDeployStatusListener listener = new HubDeployStatusListener() {
             int lastPercentageComplete = 0;
             @Override
             public void onStatusChange(int percentComplete, String message) {
                 if (percentComplete >= 0) {
                     logger.info(percentComplete + "% " + message);
                     lastPercentageComplete = percentComplete;
-                    template.convertAndSend("/topic/install-status", new StatusMessage(percentComplete, message), stompHeaders);
+                    String msg = "";
+                    if (message.endsWith("Complete")) {
+                        msg = message;
+                    } else if (isUninstall && !message.startsWith("Uninstalling")) {
+                       msg = "Uninstalling..." + message;
+                    } else {
+                       msg = "Installing..." + message;
+                    }
+                    template.convertAndSend("/topic/install-status", new StatusMessage(percentComplete, msg), stompHeaders);
                 }
             }
 
             @Override
             public void onError(String commandName, Exception exception) {
                 String message = "Error encountered running command: " + commandName;
-                template.convertAndSend("/topic/install-status", new StatusMessage(lastPercentageComplete, message));
+                String msg = "";
+                if (message.endsWith("Complete")) {
+                    msg = message;
+                } else if (isUninstall && !message.startsWith("Uninstalling")) {
+                    msg = "Uninstalling..." + message;
+                } else {
+                    msg = "Installing..." + message;
+                }
+                template.convertAndSend("/topic/install-status", new StatusMessage(lastPercentageComplete, msg));
+
                 logger.error(message, exception);
                 dataHubConfigurationException[0] = exception;
             }
         };
-        String directory = payload.get("directory").asText("");
-        // setting the project directory will resolve any relative paths
+        return listener;
+    }
+
+    private void install(String directory, HubDeployStatusListener listener, final Exception[] dataHubConfigurationException) throws Exception {
```

Comment:
I don't see any use of dataHubConfigurationException - what's its purpose? If it's not used here, can it be removed from the anonymous listener above?

Reply:
dataHubConfigurationException is based on previous install call logic, it is an excepection holder for the first thrown. We check it at the end, and rethrow it as ProjectDirectoryException. I was not sure why we did that and just did not want to break existing logic. @ryanjdew do you have any idea about
`            if (exception instanceof IOException || rootCause instanceof IOException) {
                exception = new ProjectDirectoryException("Your user account does not have write permissions to the installation directory.", "Log in as a user with write permissions to the directory you specify, or provide the absolute path to another directory.", exception);
            }`